packages/sw-precaching/src/lib/install-handler.js

```diff
@@ -13,49 +13,69 @@
  limitations under the License.
 */
 
-import assert from '../../../../lib/assert';
-import {version, defaultCacheId} from './constants';
+import ErrorFactory from './error-factory';
+import IDBHelper from '../../../../lib/idb-helper.js';
+import {version, defaultCacheId, DB_NAME, DB_VERSION, DB_STORENAME}
+  from './constants';
 
-const setOfCachedUrls = (cache) => {
-  return cache.keys()
-    .then((requests) => requests.map((request) => request.url))
-    .then((urls) => new Set(urls));
+const getPreviousRevisionDetails = (idbHelper, path) => {
+  return idbHelper.get(path);
 };
 
-const urlsToCacheKeys = (precacheConfig) => new Map(
-  /** urls.map(item => {
-    var relativeUrl = item[0];
-    var hash = item[1];
-    var absoluteUrl = new URL(relativeUrl, self.location);
-    var cacheKey = createCacheKey(absoluteUrl, hashParamName, hash, /a/);
-    return [absoluteUrl.toString(), cacheKey];
-  })**/
-);
-
-export default ({assetsAndHahes, cacheId} = {}) => {
-  assert.isType(assetsAndHahes, 'array');
-
-  self.addEventListener('install', (event) => {
-    const cacheName =
-      `${cacheId || defaultCacheId}-${version}-${self.registration.scope}`;
-
-    event.waitUntil(
-      caches.open(cacheName).then((cache) => {
-        return setOfCachedUrls(cache).then((cachedUrls) => {
-          return Promise.all(
-            Array.from(urlsToCacheKeys.values()).map(function(cacheKey) {
-              // If we don't have a key matching url in the cache already,
-              // add it.
-              if (!cachedUrls.has(cacheKey)) {
-                return cache.add(
-                  new Request(cacheKey, {credentials: 'same-origin'}));
-              }
-
-              return Promise.resolve();
-            })
-          );
-        });
-      })
-    );
+const addAssetHashToIndexedDB = (idbHelper, assetAndHash) => {
+  return idbHelper.put(assetAndHash.path, assetAndHash.revision);
+};
+
+const installHandler = ({assetsAndHahes, cacheId} = {}) => {
+  if (!Array.isArray(assetsAndHahes)) {
+    throw ErrorFactory.createError('assets-not-an-array');
+  }
+
+  const cacheName =
+    `${cacheId || defaultCacheId}-${version}-${self.registration.scope}`;
+  const idbHelper = new IDBHelper(DB_NAME, DB_VERSION, DB_STORENAME);
+
+  return caches.open(cacheName)
+  .then((openCache) => {
+    const cachePromises = assetsAndHahes.map((assetAndHash) => {
+      return getPreviousRevisionDetails(idbHelper, assetAndHash.path)
+      .then((previousRevisionDetails) => {
+        if (previousRevisionDetails) {
+          if (previousRevisionDetails === assetAndHash.revision) {
+            /* eslint-disable no-console */
+            console.log('    Already Cached? TODO: Need to check cache to ' +
+              'ensure it\'s actually already cached.');
+            /* eslint-enable no-console */
+            return Promise.resolve();
+          }
+        }
+
+        return Promise.all([
```

Comment:
Well there are two problem scenarios:

1.) Cache fails to install.
2.) The indexedDB entry fails to install.

For scenario 1 (Cache failing to install), adding to indexeddb is not a bad scenario since the library should check the cache entry exists before when it finds the manifest has defined both the same asset and revision (i.e. the module should handle this scenario).

For scenario 2, having the cached asset means that the serviceworker can still use that cache. The only difference is that on the next install, the asset will be recached when it wasn't needed (this being due to no revision data being in indexeddb).

I've switched to the cache first then add indexeddb entry since it will probably cause confusion when someone else looks at this, but it does feel like a safe thing to do in parallel.